Add JShell Accumulo Feature Blog (#274)
apache/accumulo-website: _posts/blog/2021-04-21-jshell-accumulo-feature.md

---
Title: JShell Accumulo Feature
Author: R. Dane Magbuhos

> Reviewers: Christopher Tubbs, Keith Turner

Reviewer: I'm not a huge fan of either the Author info (for the same reason Apache projects don't use the `@author` tag in source code, because it's a community-maintained group effort, that all of us are responsible for). However, I recognize that blog posts are somewhat different than source code, since they are typically authored at a point in time, rather than continuously maintained. So, I think it's fine if you want to leave that. Just know that it's optional.

I'm also not a fan of the "Reviewers" tag, either, but for additional reasons. First, it is intended to reflect reviewers of the blog post itself, not the work the blog post is describing. So, right now, it should be blank anyway. Second, I don't think it adds any value. The intent of having this seems to be to try to give credit for contributors who may have made significant contributions as editors in the review process. However, for me personally, it feels like I'm being singled out as responsible in some way for its content, rather than the entire project being responsible for its content. I'm also uncomfortable trying to weigh merit against other individuals like this (it raises the question of how much contributing makes you a reviewer, or how much contributing would make you a co-author).

So, feel free to retain your Author tag if you want, but please remove the Reviewers lines for now, and if you add it back later in response to other reviewers, please leave my name off of it, since I'm not comfortable being called out on the blog posts by name. I really wish we hadn't started this precedent in previous blog posts.

Thanks!

Author: Thank you for sharing your perspective on the tags. I agree the blog posts should just inform the users about new features and additions to Accumulo and not about who worked on what. At the time of posting, I was following the format from previous blogs and was unaware about this. I'll remove the author and reviewer tags since they are not needed at all.

Reviewer: To be clear, this is just my perspective. Other members of the community may have a different perspective. You are welcome to include them at your discretion (just not with my name in the reviewers).

---

## Overview

> First introduced in Java 9+, [JShell][JShell Doc] is an interactive Read-Evaluate-Print-Loop (REPL) 

Reviewer: ```suggestion
First introduced in Java 9, [JShell][JShell Doc] is an interactive Read-Evaluate-Print-Loop (REPL) 
```

> Java tool that assess user's inputed declarations, statements, and expressions and outputs 

Reviewer: ```suggestion
Java tool that interprets user's input as Java and outputs 
```

the results. This tool provides a convenient way to test out and execute quick tasks with Accumulo
in the terminal.

## Major Features
* During Accumulo build, produces a default JShell script called `jshell-init.jsh` 
containing up-to-date [Accumulo Java APIs][public APIs] and [AccumuloClient][client] 

> build implementation

Reviewer: Instead of mentioning how it is built, which isn't quite relevant to end users, it should instead mention that a default `conf/jshell-init.jsh` file is provided in our binary distribution tarball, and includes some initial imports for interacting with Accumulo's API.

It might be worth mentioning that JShell could be run manually using the user's installed `jshell` command from their installation of Java, but that by running it with `bin/accumulo jshell`, it will automatically pick up the `CLASSPATH` and other environment from the user's `conf/accumulo-env.sh` and automatically initialize with the provided `conf/jshell-init.jsh` file, which can also be customized by users.

Since all of this is something a user could already do themselves, I think the main focus of this blog post should be on the convenience that it offers to users, so that they don't have to do it all manually.

* Startup JShell with default or custom JShell script 

> * Both JShell start up options automatically import all relevant Java APIs 

Reviewer: By default they import classes to interact with the Accumulo API... but might not be all relevant Java APIs for the user's intended task. Also, this is only by default. The file is customizable by the user in their conf/ directory.

## Booting Up JShell Accumulo

> After installing and configuring the [latest Accumulo update][accumulo-repo]

Reviewer: Since this is a 2.1 feature, we could point to the downloads page. If we want to publish this blog post before 2.1 is released, then some of the wording could be tailored to refer to the "upcoming 2.1 release" for this feature, with perhaps a minor mention that they could build pre-release snapshot versions to try it out from the git repository if 2.1 isn't released by the time they read this.

Author: I believe we can post this blog before the 2.1 release. That way users can get a small glimpse of what 2.1 has to offer.

Reviewer: I agree. However, we do need to be careful that we're not violating ASF policy regarding publication of non-released material: https://www.apache.org/legal/release-policy.html#publication

Specifically, we want to avoid directing users (non-devs for the project) to unreleased content.

Author: Quick question: Since JShell Accumulo has been merged with the main Accumulo repo, would it be safe to link the main Accumulo repo on the blog post?

Reviewer: The git repository hosting Accumulo source code, and especially the branch where this feature is merged, contains unreleased content. See the definition of "release" at ASF at https://www.apache.org/legal/release-policy.html#release-definition

I advise not linking against it. We can link to the downloads page for "when it becomes available in 2.1", and we can make a reference to the unreleased code in the git repository, but we probably shouldn't link to it. Developers who are involved already understand how to access unreleased code, so don't need to be told, and users who aren't involved in the development shouldn't be directed to use any unreleased code.

So, I think one possible way to frame this is with the following elements:
1. Explanation of upcoming feature
2. If you're a developer who wants to get involved in testing, contact us (link to our contact-us page) or see our contributing guide (link to contributing docs)
3. Link to download page with some kind of statement like "look for this feature in an upcoming 2.1 release at our downloads page" or something along those lines.

Author: Ahhh I see. That makes more sense. I'll get rid of the repo link and will implement your suggested idea.

follow the steps below to startup JShell:

> 1) Open up a terminal and navigate to Accumulo's home directory 

Reviewer: Since "home directory" has a special meaning in Linux environments, it'd be better to refer to the Accumulo installation directory, if using the binary distribution tarball we provide. If they're using a distribution from a commercial vendor or other distributor, their experience may be different.

2) To startup JShell with **default script** use this command:

```bash
$ bin/accumulo jshell 
```
3) To startup JShell with **custom script** use this command:

```bash
$ bin/accumulo jshell --startup (file/path/to/custom_script.jsh)
```
## JShell Accumulo Default Script
The auto-generated `jshell-init.jsh` is located in Accumulo's `conf/` directory. 
Inside `jshell-init.jsh` contains [Accumulo Java APIs][public APIs] formatted as import statements 
and [AccumuloClient][client] build implementation. On startup the script automatically loads in the 
APIs and attempts to construct a client. Should additional APIs and/or code implementations be 
needed, simply append them to `jshell-init.jsh`. Alternatively you can create a separate JShell 

> script and specify the custom script's file path on startup.

Reviewer: Might be worth mentioning that you can append any options that you might append to the `jshell` command-line if you were running that directly.

|  Welcome to JShell -- Version 11.0.10
|  For an introduction type: /help intro

jshell> 
```

## JShell Accumulo Example
1) Booting up JShell using default script

```
Preparing JShell for Apache Accumulo 

Building Accumulo client using 'file:/home/accumulo/conf/accumulo-client.properties'

Use 'client' to interact with Accumulo

|  Welcome to JShell -- Version 11.0.10
|  For an introduction type: /help intro

jshell> 
```

2) Providing JShell with an Accumulo task

```Java
  // Create a table called "GothamPD".
  client.tableOperations().create("GothamPD");

  // Create a Mutation object to hold all changes to a row in a table. 
  // Each row has a unique row ID.
  Mutation mutation = new Mutation("id0001");

  // Create key/value pairs for Batman. Put them in the "hero" family.
  mutation.put("hero", "alias", "Batman");
  mutation.put("hero", "name", "Bruce Wayne");
  mutation.put("hero", "wearsCape?", "true");

  // Create a BatchWriter to the GothamPD table and add your mutation to it. 
  // Try w/ resources will close for us.
  try (BatchWriter writer = client.createBatchWriter("GothamPD")) {
      writer.addMutation(mutation);
  }

>   // Read and print all rows of the "GothamPD" table. 
>   // Try w/ resources will close for us.
>   try (org.apache.accumulo.core.client.Scanner scan =
>     client.createScanner("GothamPD", Authorizations.EMPTY)) {

Reviewer: Might be worth a mention that the fully-qualified class name for Scanner needs to be used because of a conflict with Java's built-in `java.util.Scanner`, but that if they want to use a shortname, they can assign it to the `ScannerBase` type instead.

```suggestion
  // Read and print all rows of the "GothamPD" table. 
  // Try w/ resources will close for us.
  try (ScannerBase scan = client.createScanner("GothamPD", Authorizations.EMPTY)) {
```

    System.out.println("Gotham Police Department Persons of Interest:");
    

>     // A Scanner is an extension of java.lang.Iterable so behaves just like one.
>     for (Map.Entry<Key,Value> entry : scan) {
>       System.out.printf("Key : %-50s  Value : %s\n", entry.getKey(), entry.getValue());
>     }

Reviewer: As of 2.1, we can use scanner.forEach, which makes this slightly nicer example code:

```suggestion
    scan.forEach((k, v) -> System.out.printf("Key : %-50s  Value : %s\n", k, v));
```

---
Title: JShell Accumulo Feature

---

## Overview
First introduced in Java 9, [JShell][jshell-doc] is an interactive Read-Evaluate-Print-Loop (REPL) 
Java tool that interprets user's input and outputs the results. This tool provides a convenient way 

> to test out and execute quick tasks with Accumulo in the terminal. This feature is apart of the 

Reviewer: ```suggestion
to test out and execute quick tasks with Accumulo in the terminal. This feature is a part of the 
```

upcoming Accumulo 2.1 release. If you're a developer and want to get involved in testing, 

> [contact us][contact] or review our [contributing guide][guide]

Reviewer: ```suggestion
[contact us][contact] or review our [contributing guide][guide].
```

## Major Features
* Default JShell script provides initial imports for interacting with Accumulo's API and 
provided in Accumulo's binary distribution tarball

* On startup, JShell Accumulo  will automatically import the `CLASSPATH`, load in a configured 
environment from user's `conf/accumulo-env.sh`, and invoke `conf/jshell-init.jsh` to 
allow rapid Accumulo task executions

* JShell Accumulo can startup using default/custom JShell script and users can append any JShell 
command-line [options][jshell-option] to the startup command

## Booting Up JShell Accumulo
1) Open up a terminal and navigate to Accumulo's installation directory 

2) To startup JShell with **default script** use this command:

```bash
$ bin/accumulo jshell 
```
3) To startup JShell with **custom script** use this command:

```bash

> $ bin/accumulo jshell --startup (file/path/to/custom_script.jsh)

Reviewer: ```suggestion
$ bin/accumulo jshell --startup file/path/to/custom_script.jsh
```

```

> **Note:** Executing this specific command  _"$jshell"_  will startup JShell, however you will be required to manually import the `CLASSPATH` and the configured environment from `conf/accumulo-env.sh`. In addition, you will need to provide the file directory path to `conf/jshell-init.jsh` before any Accumulo tasks can be performed. Using one of the startup commands above will automate 
> the setup process.

Reviewer: ```suggestion
**Note:** One can execute the `jshell`  command to startup JShell. However, doing so will
require manually importing the `CLASSPATH` and the configured environment from
`conf/accumulo-env.sh` and manually specifying the startup file for `conf/jshell-init.jsh` before
any Accumulo tasks can be performed. Using one of the startup commands above will
automate that process for convenience.
```

## JShell Accumulo Default Script

> The auto-generated `jshell-init.jsh` is a customizable file located in Accumulo's installation `conf/` directory. Inside `jshell-init.jsh` contains [Accumulo Java APIs][public APIs] formatted as import statements and [AccumuloClient][client] build implementation. On startup the script automatically loads in the APIs and attempts to construct a client. Should additional APIs and/or code implementations be needed, simply append them to `jshell-init.jsh`. Alternatively you can create a separate JShell script and specify the custom script's file path on startup.

Reviewer: Adding a few commas after some transition words:

```suggestion
The auto-generated `jshell-init.jsh` is a customizable file located in Accumulo's installation `conf/` directory. Inside, `jshell-init.jsh` contains [Accumulo Java APIs][public APIs] formatted as import statements and [AccumuloClient][client] build implementation. On startup, the script automatically loads in the APIs and attempts to construct a client. Should additional APIs and/or code implementations be needed, simply append them to `jshell-init.jsh`. Alternatively, you can create a separate JShell script and specify the custom script's file path on startup.
```

Also, it might be nice to wrap these long lines. I tend to use `vim` to edit files, and generally use `:set tw=65` or then use visual mode to select and `gq` to wrap. I might have to tweak to avoid wrapping specially-formatted markdown, but this works for the most part. It's not strictly necessary to wrap these for Markdown, but it makes looking at the source a little easier sometimes.

Author: Sorry about that. I was using the Eclipse IDE when creating this blog post. I'll use vim to word wrap the sentences.

Reviewer: It's fine. Use whatever you prefer. I don't know if Eclipse will automatically wrap/format Markdown content or plain text files, but I know it's possible to configure it to show a vertical line at a specific line length, so if you exceed it, you can wrap manually.

> The build implementation finds and uses `accumulo-client.properties` in Accumulo's 
> classpath to auto-generate an [AccumuloClient][client] called **client**. 

Reviewer: It's not clear what "The build implementation" refers to.

```suggestion
To construct an [AccumuloClient][client], the provided `conf/jshell-init.sh` script finds and uses `accumulo-client.properties` on Accumulo's class path, and assigns the result to a variable called **client**.
```
